src/v/cluster/scheduling/allocation_strategy.cc

```diff
@@ -95,13 +99,22 @@ model::node_id find_best_fit(
             const allocation_constraints::soft_constraint_ev_ptr& ev) {
               return score + ev->score(*node);
           });
-        if (score > last_score) {
-            last_score = score;
-            best_fit = it->first;
+
+        if (score >= best_score) {
+            if (score > best_score) {
+                // untied, winner clear out existing winners
+                best_score = score;
+                best_fits.clear();
+            }
+            best_fits.push_back(it->first);
         }
     }
 
-    return best_fit;
+    vassert(!best_fits.empty(), "best_fits empty");
+
+    // we break ties randomly, by selecting a random node out of those
+    // with the highest score
+    return best_fits.at(random_generators::get_int(best_fits.size() - 1));
```

Comment:
This is the substance of this change: we break ties randomly after collecting all the tied-for-best candidates.

Reply:
You already verify size. Can use bracket operator. Good idea here.

Reply:
> You already verify size.

Are you referring to the assert? This assets a different thing: that we have at least one "best_fit" element, which is different than the earlier check of at least one possible_node. ISTM that the latter will always lead to the former, but that's not a proof (e.g., there is at least one place where we `continue` in the loop w/o a chance to add to the best_fits vector).

Reply:
I this this was referring to avoid the extra bounds checking contained in vector::at. but I wouldn't consider random_generators::get_int to have a particularly strong contract